cirq/sim/mux_test.py

```diff
@@ -238,3 +238,125 @@ def test_repetitions(repetitions):
+
+    np.testing.assert_allclose(cirq.final_density_matrix(cirq.X,
+                                                         initial_state=1),
+                               [[1, 0], [0, 0]],
+                               atol=1e-8)
+
+    np.testing.assert_allclose(cirq.final_density_matrix(
+        cirq.X, initial_state=[np.sqrt(0.5), 1j * np.sqrt(0.5)]),
+                               [[0.5, 0.5j], [-0.5j, 0.5]],
+                               atol=1e-8)
+
+
+def test_final_density_matrix_dtype_insensitive_to_initial_state():
+    assert cirq.final_density_matrix(cirq.X,).dtype == np.complex64
+
+    assert cirq.final_density_matrix(cirq.X,
+                                     initial_state=0).dtype == np.complex64
+
+    assert cirq.final_density_matrix(cirq.X,
+                                     initial_state=[np.sqrt(0.5),
+                                                    np.sqrt(0.5)
+                                                   ]).dtype == np.complex64
+
+    assert cirq.final_density_matrix(cirq.X,
+                                     initial_state=np.array(
+                                         [np.sqrt(0.5),
+                                          np.sqrt(0.5)])).dtype == np.complex64
+
+    for t in [np.int32, np.float32, np.float64, np.complex64]:
+        assert cirq.final_density_matrix(
+            cirq.X, initial_state=np.array([1, 0],
+                                           dtype=t)).dtype == np.complex64
+
+        assert cirq.final_density_matrix(
+            cirq.X,
+            initial_state=np.array([1, 0], dtype=t),
+            dtype=np.complex128).dtype == np.complex128
+
+
+def test_final_density_matrix_param_resolver():
+    s = sympy.Symbol('s')
+
+    with pytest.raises(ValueError, match='not specified in parameter sweep'):
+        _ = cirq.final_density_matrix(cirq.X**s)
+
+    np.testing.assert_allclose(
+        cirq.final_density_matrix(cirq.X**s, param_resolver={s: 0.5}),
+        [[0.5 - 0.j, 0. + 0.5j], [0. - 0.5j, 0.5 - 0.j]])
+
+
+def test_final_density_matrix_qubit_order():
+    a, b = cirq.LineQubit.range(2)
+
+    np.testing.assert_allclose(
+        cirq.final_density_matrix([cirq.X(a), cirq.X(b)**0.5],
+                                  qubit_order=[a, b]),
+        [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0.5, 0.5j], [0, 0, -0.5j, 0.5]])
+
+    np.testing.assert_allclose(
+        cirq.final_density_matrix([cirq.X(a), cirq.X(b)**0.5],
+                                  qubit_order=[b, a]),
+        [[0, 0, 0, 0], [0, 0.5, 0, 0.5j], [0, 0, 0, 0], [0, -0.5j, 0, 0.5]])
+
+    np.testing.assert_allclose(
+        cirq.final_density_matrix([cirq.X(a), cirq.X(b)**0.5],
+                                  qubit_order=[b, a],
+                                  noise=cirq.ConstantQubitNoiseModel(
+                                      cirq.amplitude_damp(1.0))),
+        [[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
+
+
+def test_final_density_matrix_seed():
```

Comment:
Hmmmm. This is a bit of a curve ball, but I think that `final_density_matrix` should treat measurement as dephasing instead of as a collapsing process. That is to say, I think the final density matrix of the circuit "H then measure" should be the maximally mixed state instead of randomly |0> or |1>.

There should perhaps be an argument to the method to decide between the two behaviors. The basic issue is that "final density matrix" should be a single thing, instead of a distribution of things. We should probably discuss this at the sync tomorrow, since it's a rather big semantic difference.

Reply:
I made the changes you have suggested and added a few tests (probably not enough).

Reply:
Done.